src/coreclr/vm/jitinterface.cpp

```diff
@@ -2109,15 +2109,19 @@ static GetTypeLayoutResult GetTypeLayoutHelper(
     parNode.hasSignificantPadding = false;
 
     EEClass* pClass = pMT->GetClass();
-    if (pClass->IsNotTightlyPacked() && (!pClass->IsManagedSequential() || pClass->HasExplicitSize() || pClass->IsInlineArray()))
+    if (pClass->IsNotTightlyPacked())
     {
-        // Historically on the JIT side we did not consider types with GC
-        // pointers to have significant padding, even when they have explicit
-        // layout attributes. This retains the more liberal treatment and
-        // lets the JIT still optimize these cases.
-        if (!pMT->ContainsPointers() && pMT != g_TypedReferenceMT)
+        if (pClass->HasExplicitFieldOffsetLayout() || (pClass->IsManagedSequential() && pClass->HasExplicitSize()) ||
```

Comment:
```suggestion
        if (pClass->HasExplicitFieldOffsetLayout() || pClass->HasExplicitSize() ||
```
Can it be simplified like this?

Reply:
Good question. Let me try to check that.

Reply:
I think we can in the VM, we only set `HasExplicitSize` for sequential/explicit structs in the first place, size specifications for auto-layout structs are ignored. However I don't think we would be able to make a similar change on the crossgen2 side as it seems to more directly look at the metadata. So I think I will leave it with the explicit check for sequential layout in both places.

Reply:
The condition does not exactly match between VM and crossgen currently (the VM checks IsManagedSequential, crossgen2 checks IsSequentialLayout, etc.). Are you going to make them match exactly? Otherwise, I think it is ok for the exact condition to differ - as long as it produces the same result.

Reply:
No, but is very subtle and not at all obvious that `HasExplicitSize()` returns false for auto-layout types with explicit size in metadata, and not immediately obvious whether the missing check is a bug or not if you compare the VM to crossgen2 version.
But I can switch and add comment on `HasExplicitSize` to this regard. Do you think that'd be better?

Reply:
I like that `pClass->HasExplicitFieldOffsetLayout() || pClass->HasExplicitSize()` clearly communicates the intent: if the offsets or size are explicitly specified by the user (and respected by the runtime), then do this.

If there is something that can use comment, I think it would be the crossgen implementation. It is much harder to see the intent from it.

(I do not have a strong opinion about any of this.)

Reply:
Makes sense to me, I've applied your suggestion.